File: zhill/bot.py
import re
import traceback
import urllib.error
import urllib.request
from datetime import datetime

import irc.bot

class Bot(irc.bot.SingleServerIRCBot):
    def __init__(self, hill):
        self._hill = hill
        self._cfg = hill.cfg()['irc']
        self._command = self._cfg['command']
        self._log = open('hill.log', 'a')

        cfg = self._cfg
        irc.bot.SingleServerIRCBot.__init__(self, [(cfg['server'], cfg.get('port', 6667))], cfg['nick'], cfg['nick'])

    def on_nicknameinuse(self, c, e):
        c.nick(c.get_nickname() + '_')

    def on_welcome(self, c, e):
        for ch in self._hill.cfg()['irc']['channels'].split(' '):
            c.join(ch)

    def on_pubmsg(self, c, e):
        msg = e.arguments[0].split(' ', 3)
        if len(msg) > 0 and msg[0] == self._command:
            self._handle(c, e.target, e.source.nick + ': ', e.source.nick, msg)

    def on_privmsg(self, c, e):
        msg = e.arguments[0].split(' ', 3)
        if len(msg) > 0 and msg[0] == self._command:
            self._handle(c, e.source.nick, '', e.source.nick, msg)

    def _handle(self, c, target, prefix, nick, msg):
        if len(msg) <= 2:
            c.privmsg(target, prefix + '"!bfjoust progname code". See http://zem.fi/bfjoust/ for documentation.')
            return

        orig_prog = msg[1]
        code = msg[2]

        prog = re.sub(r'[^a-zA-Z0-9_-]', '', orig_prog)
        if len(prog) < 0.75*len(orig_prog):
            c.privmsg(target, prefix + 'Program name looks like gibberish. Did you forget it?')
            return

        prog = nick + '~' + prog

        tstamp = datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')
        self._log.write('{0} {1} {2}\n'.format(tstamp, nick, code))
        self._log.flush()

        if re.match(r'(?:https?|ftp)://', code):
            err, code = fetch(code)
            if err is not None:
                c.privmsg(target, prefix + 'URL fetch problems: ' + err)
                return
        else:
            code = code.encode('utf-8')

        try:
            # run challenge

            err, result = self._hill.challenge(prog, code)
            if err is not None:
                c.privmsg(target, err)
                return
            c.privmsg(target, result)

            # update web components

            report = self._cfg.get('report')
            if report is not None:
                with open(report, 'w') as f:
                    self._hill.write_report(f)

            breakdown = self._cfg.get('breakdown')
            if breakdown is not None:
                with open(breakdown, 'w') as f:
                    self._hill.write_breakdown(prog, f)

        except Exception:
            traceback.print_exc(file=self._log)
            c.privmsg(target, 'I broke down! Ask fizzie to help! The details are in the log!')
            self.die()

def fetch(url):
    MAXSIZE = 2*1024*1024

    try:
        with urllib.request.urlopen(url, timeout=10) as resp:
            if resp.getcode() != 200:
                return 'request failed: ' + resp.getcode(), None
            code = resp.read(MAXSIZE+1)
            if len(code) > MAXSIZE:
                return 'maximum size limit ({0} bytes) exceeded', None
            return None, code
    except urllib.error.URLError as e:
        return e.reason, None

File: zhill/bot.rb
require 'cinch'

require_relative 'gear'

# IRC bot to run a public gearlanced hill.
#
# Based on the Cinch[https://github.com/cinchrb/cinch] framework.

module Bot
  # Constructs a new bot.
  #
  # +cfg+ should be a hash containing the following (string) keys.
  #
  # [nick]            Nickname for the bot.
  # [server]          IRC server to connect to.
  # [port]            (opt) IRC server port, 6667 by default.
  # [channels]        Space-separated list of IRC channels to join.
  # [summarychannel]  (opt) Channel to broadcast all hill-changing
  #                   summary messages to.
  # [command]         Command for submitting, e.g. "!bfjoust".
  # [testcommand]     (opt) Command for testing a submission
  #                   without changing the hill.
  # [help]            (opt) Command to print the (fixed) help text.
  #
  # Normally +cfg+ is the +irc+ section from the YAML configuration
  # file for a HillManager.
  #
  # +server+ must be a Server object serving the target hill.
  def Bot.make(cfg, server)
    command = cfg['command']
    testcommand = cfg['testcommand']

    Cinch::Bot.new do
      configure do |c|
        c.server = cfg['server']
        c.port = (cfg['port'] || '6667').to_i
        c.nick = cfg['nick']
        c.realname = cfg['nick']
        c.channels = cfg['channels'].split
      end

      # help command

      if cfg['help']
        on :message, cfg['help'] do |m|
          m.reply('I do !bfjoust; see http://zem.fi/bfjoust/ for more information.', true)
        end
      end

      # register summary channel as hill server's herald

      if cfg['summarychannel']
        server.herald do |message|
          Channel(cfg['summarychannel']).msg(message)
        end
      end

      # do or try a BF Joust match

      joust = lambda do |m, orig_prog, code, try_only|
        prog = orig_prog.gsub(/[^a-zA-Z0-9_-]/, '')
        if prog != orig_prog
          m.reply("Program name (#{orig_prog}) is restricted to characters in [a-zA-Z0-9_-], sorry.", true)
          break
        elsif prog.length > 48
          m.reply('Program name is overly long. 48 characters should be enough for everyone.', true)
          break
        end

        nick = m.user.nick
        prog = nick + '.' + prog

        server.log(prog + ' ' + code)
        server.joust(prog, code, try_only) do |result, message|
          if result == :fatal
            m.reply('I broke down! Ask fizzie to help! The details are in the log! ' + message, true)
            m.bot.quit('Abandon ship, abandon ship!')
          end

          m.reply(message, result == :error)
          if result == :ok &&
              cfg['summarychannel'] &&
              m.target.name != cfg['summarychannel'] &&
              !try_only
            Channel(cfg['summarychannel']).msg(message)
          end
        end
      end

      # handle the main command

      on :message, /^#{Regexp.escape(command)}/ do |m|
        msg = m.message.split(' ', 3)

        if msg.length <= 2
          m.reply("\"#{command} progname code\". See http://zem.fi/bfjoust/ for documentation.", true)
          break
        end

        joust.call(m, msg[1], msg[2], false)
      end

      # handle the test command

      unless testcommand.nil?
        on :message, /^#{Regexp.escape(testcommand)}/ do |m|
          msg = m.message.split(' ', 3)

          if msg.length <= 2
            m.reply("\"#{testcommand} progname code\". See http://zem.fi/bfjoust/ for documentation.", true)
            break
          end

          joust.call(m, msg[1], msg[2], true)
        end
      end
    end
  end
end
